src/app/components/routemap/RouteMap.tsx

```diff
@@ -0,0 +1,380 @@
+            if(refr?.childNodes.length > 1){
+                refr?.removeChild(refr?.lastChild as Node)
+            }
+        }
+    }
+
+    const timetableType = (type: string, index:number) => {
+        if(type === 'C'){
+            return {ref: refcyc, index: index, chipColor: 'bg-chip-red'}
+        } else if(type === 'DHJ'){
+            return {ref: refjun, index: index, chipColor: 'bg-chip-purple'}
+        } else if(type === 'DY'){
+            return {ref: refyes, index: index, chipColor: 'bg-chip-green'}
+        } else {
+            return {ref: refdir, index: index, chipColor: 'bg-chip-blue'}
+        }
+    }
+
+    useEffect(() => {
+        fetchLanguages()
+    },[i18n.language, fetchLanguages])
+
+    useEffect(() => {
+        directInput()
+        directLineInput()
+        window.addEventListener("resize",updateLines)
+        return () => {
+            window.removeEventListener("resize", updateLines)
+        }
+    }, [updateLines, directInput, directLineInput])
+
+    useEffect(() => {
+        const circleAnimationRemoveAll = () => {
+            circleAnimationRemove(refdir)
+            circleAnimationRemove(refcyc)
+            circleAnimationRemove(refyes)
+            circleAnimationRemove(refjun)
+        }
+        const updateHighlight = () => {
+            if(linecyc.current.length > 0 && linedir.current.length > 0 && lineyes.current.length > 0 && linejun.current.length > 0){
+                circleAnimationRemoveAll()
+                if(timetable?.time !== ''){
+                    if(props.tab === 'shuttlecoke_o'){
+                        circleAnimation(timetableType(timetable.type,1))
+                    } else if(props.tab === 'subway'){
+                        circleAnimation(timetableType(timetable.type,2))
+                    } else if(props.tab === 'yesulin'){
+                        circleAnimation(timetableType(timetable.type,3))
+                    } else if(props.tab === 'jungang'){
+                        circleAnimation({ref: refjun, index: 3, chipColor: 'bg-chip-purple'})
+                    } else if(props.tab === 'shuttlecoke_i'){
+                        if(timetable.type === 'NA'){
+                            return
+                        }
+                        circleAnimation({ref: refdir, index: 3, chipColor: 'bg-chip-blue'})
+                        circleAnimation({ref: refjun, index: 4, chipColor: 'bg-chip-purple'})
+                        circleAnimation({ref: refcyc, index: 4, chipColor: 'bg-chip-red'})
+                        circleAnimation({ref: refyes, index: 4, chipColor: 'bg-chip-green'})
+                    } else {
+                        circleAnimation(timetableType(timetable.type,0))
+                    }
+                }
+            }
+        }
+        updateLines()
+        updateHighlight()
+    },[direct, cycle, yesulin, jungang, dirLine, cycLine, yesLine, junLine, props.tab, timetable, updateLines])
```

Comment:
The `useEffect` is doing to much of its own.

`useEffect` is executed right after the Virtual DOM is created. Consider carefully if this logic inside useEffect really needs to be executed at that moment.

You might not need useEffect that much.